sycamore/transforms/sketcher.py

```diff
@@ -0,0 +1,124 @@
+    """
+    For each Document, uses shingling to hash sliding windows of the
+    text_representation using various permutations.  Uses SimHash
+    to reduce each shingle to a similarity hash.  The set of SimHashes
+    is called the sketch.  Documents' sketches can be compared to
+    determine if they have near-duplicate content.  The SketchUniquify
+    transform can be used to de-duplicate small docsets in Sycamore.
+    De-duplicating at retrieval-time is more scalable and avoids some
+    relevance problems.
+
+    Args:
+        child: The source node or component that provides the documents
+        window: Number of bytes in the sliding window that is hashed (32)
+        courses: Number of hashes comprising a shingle (15, must be odd)
+        tabs: Number of permutation variants in each shingle (8)
+
+    Example:
+        .. code-block:: python
+
+            node = ...  # source node or component that provides hierarchical documents.
+            xform = Sketcher(child=node)
+            dataset = xform.execute()
+    """
+
+    def __init__(self, child: Node, window: int = 32, courses: int = 15, tabs: int = 8, **kwargs):
+        super().__init__(child, **kwargs)
+        self.window = window
+        self.courses = courses
+        self.tabs = tabs
+
+    class Callable:
+        def __init__(self, window: int, courses: int, tabs: int):
+            self.window = window
+            self.courses = courses
+            self.tabs = tabs
+
+        def run(self, doc: Document) -> Document:
+            txt = doc.text_representation
+            if txt:
+                utf = normalizeString(txt).encode("utf-8")
+                doc.simHashes = simHashText(utf, self.window, self.courses, self.tabs)
+            return doc
+
+    def execute(self) -> Dataset:
+        dataset = self.child().execute()
+        xform = Sketcher.Callable(self.window, self.courses, self.tabs)
+        return dataset.map(generate_map_function(xform.run))
+
+
+class SketchUniquify(SingleThreadUser, NonGPUUser, Transform):
+    """
+    Removes each Document which is a near-duplicate of a Document seen
+    before.  Uses the SimHash values calculated by the Sketcher transform.
+    This approach requires full materialization of the entire docset on a
+    single node.  It will store all sketches in memory.  It is not
+    suitable for large docsets.
+
+    Args:
+        child: The source node or component that provides the documents
+        threshold: Largest distance to be considered a duplicate (16)
+
+    Example:
+        .. code-block:: python
+
+           node = ...  # source node
+           xform = SketchUniquify(child=node)
+           dataset = xform.execute()
+    """
+
+    def __init__(self, child: Node, threshold: float = 16, **kwargs):
+        super().__init__(child, **kwargs)
+        self.threshold = threshold
+
+    def execute(self) -> Dataset:
+        ds = self.child().execute()
+        ds = ds.materialize()
```

Comment:
Doesn't this materialize on multiple nodes (in the object store)? Guess I was expecting a take_all.

Reply:
Thinking about this a bit more, I'm trying to compare `materialize()` and then `iter_rows()` with just doing a `take_call`. I guess the former materializes in the aggregate memory of the cluster and then iterates over the rows on a single node, so theoretically you don't need the whole thing in memory. But since you are still storing all the simHashes in memory it might not matter much?

Reply:
I'm no Ray expert, but I think I agree with your later comment.  The important thing is that what I'm storing in memory is a tiny subset of each Document.  Reading the doc, it seems that take_all() is more likely to OOM the local host, but that's speculation.